src/combinatorics/simple_graph/ends/defs.lean

```diff
@@ -0,0 +1,233 @@
+/-
+Copyright (c) 2022 Anand Rao, Rémi Bottinelli. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Anand Rao, Rémi Bottinelli
+-/
+import combinatorics.simple_graph.basic
+import combinatorics.simple_graph.connectivity
+import data.set_like.basic
+import category_theory.category.basic
+import category_theory.filtered
+import topology.category.Top.limits
+
+/-!
+# Ends
+
+This file contains a definition of the ends of a simple graph, as sections of the inverse system
+assigning, to each finite set of vertices, the connected components of its complement.
+-/
+
+universes u
+variables {V : Type u} (G : simple_graph V) (K L L' M : set V)
+
+open classical
+
+noncomputable theory
+local attribute [instance] prop_decidable
+
+namespace simple_graph
+
+section out
+
+/-- The graph induced by removing `K` -/
+@[reducible] def out := G.induce $ Kᶜ
+
+/-- Subsetship induces an obvious map on the induced graphs. -/
+@[reducible] def out_hom {K L} (h : K ⊆ L) : G.out L →g G.out K :=
+{ to_fun := λ ⟨v, hvK⟩, ⟨v, set.compl_subset_compl.mpr h hvK⟩,
```

Comment:
A warning: using patterns like `⟨v, hvK⟩` in lambdas can lead to strange difficulties when you prove things. It tends to be better to instead write this as
```suggestion
{ to_fun := λ v', ⟨v', set.compl_subset_compl.mpr h v'.property⟩,
```
(I'm relying on Lean coercing the `v'` here. You can put an explicit coercion arrow there if you want.)

Reply:
Thanks, I wasn't aware of this. We had used the `.property` syntax earlier, but I thought that pattern matching might lead to shorter proofs. I'll make this and other downstream fixes that arise soon.